8308094: Add a compilation timeout flag to catch long running compilations

mhaessig opened the pull request:
This PR adds `-XX:CompileTaskTimeout` on Linux to limit the amount of time a compilation task can run. The goal of this is initially to be able to find and investigate long-running compilations.

The timeout is implemented using a POSIX timer that sends a `SIGALRM` to the compiler thread the compile task is running on. Each compiler thread registers a signal handler that triggers an assert upon receiving `SIGALRM`. This is currently only implemented for Linux, because it relies on `SIGEV_THREAD_ID` to get the signal delivered to the same thread that timed out.

Since `SIGALRM` is now used, the test `runtime/signal/TestSigalrm.java` now requires `vm.flagless` so it will not interfere with the compiler thread signal handlers.

Testing:
 - [x] Github Actions
 - [x] tier1, tier2 on all platforms
 - [x] tier3, tier4 and Oracle internal testing on Linux fastdebug
 - [x] tier1 through tier4 with `-XX:CompileTaskTimeout=60000` (one minute timeout) to see what fails (`compiler/codegen/TestAntiDependenciesHighMemUsage2.java`, `compiler/loopopts/TestMaxLoopOptsCountReached.java`, and `compiler/c2/TestScalarReplacementMaxLiveNodes.java` fail)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308094](https://bugs.openjdk.org/browse/JDK-8308094): Add a compilation timeout flag to catch long running compilations (**Enhancement** - P4)


### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Contributors
 * Dean Long `<dlong@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26023/head:pull/26023` \
`$ git checkout pull/26023`

Update a local copy of the PR: \
`$ git checkout pull/26023` \
`$ git pull https://git.openjdk.org/jdk.git pull/26023/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26023`

View PR using the GUI difftool: \
`$ git pr show -t 26023`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26023.diff">https://git.openjdk.org/jdk/pull/26023.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26023#issuecomment-3013954027)
</details>

mhaessig commented:
/touch

dean-long commented:
This looks correct, but would it be possible to move the Linux-specific code out of src/hotspot/share?

mhaessig commented on src/hotspot/share/compiler/compilerThread.cpp:
```diff
@@ -71,3 +85,90 @@ void CompilerThread::thread_entry(JavaThread* thread, TRAPS) {
+  void CompilerThread::compiler_signal_handler(int signo, siginfo_t* info, void* context) {
+    switch (signo) {
+      case TIMEOUT_SIGNAL: {
+        assert(!Atomic::load_acquire(&_timeout_armed), "compile task timed out");
```
The compiler thread setting and unsetting the flag and the signal handler reading the flag are racing each other as soon as the timer is set, since signals are preemptive. This prevents a few false positive timeouts on architectures with weak memory models, but does not have any effect on x86 for example.

dean-long replied:
The signal is delivered to the same compiler thread, so I don't think acquire and release help here.  There shouldn't be a weak memory model issue between a thread and it's signal handler on the same thread.  Can you explain the sequence of events that could cause a false positive?

dholmes-ora replied:
That is not an accurate characterisation IMO - a memory model defines the consistency of the views of memory by different threads of execution. Any given thread of execution must always have a self-consistent view of memory, and neither the compiler or the CPU is allowed to violate that if code is actually going to be guaranteed to work correctly. If the signal is being handled in the same thread that is setting the flag then you cannot have a memory ordering issue.

And acquire/release is not the right tool just to establish an ordering: you would generally want a storestore barrier on the writer side, and a loadload barrier on the reader side.

mhaessig commented:
Thank you for reviewing this PR, @dean-long.

For v2 I factored the Linux-specific code out of hotspot/share, moved the disarm above the if, and removed the acquire/release.

dean-long commented:
Thinking about _timeout_armed a little more, the fact the the signal handler received TIMEOUT_SIGNAL should be enough.  The value of _timeout_armed should be redundant, and your assert could be changed to:

assert(false, "compile task timed out");

and _timeout_armed could be removed. It's just an inexact mirror of the timer state.

chhagedorn commented on src/hotspot/os/linux/compilerThreadTimeout_linux.cpp:
```diff
@@ -0,0 +1,124 @@
+void CompilerThreadTimeoutLinux::compiler_signal_handler(int signo, siginfo_t* info, void* context) {
+  switch (signo) {
+    case TIMEOUT_SIGNAL: {
+      assert(false, "compile task timed out");
```
Can you somehow also print the task which caused the timeout? Will just accessing `CompilerThread::current()->task()` work?

chhagedorn commented on src/hotspot/os/linux/compilerThreadTimeout_linux.hpp:
```diff
@@ -0,0 +1,53 @@
+#endif // ASSERT
+#endif // !PRODUCT
+ public:
+  CompilerThreadTimeoutLinux() NOT_PRODUCT(DEBUG_ONLY(: _timer(nullptr))) {};
```
Why do you need the `NOT_PRODUCT`? It only wraps `DEBUG_ONLY`. If that's not set, the `NOT_PRODUCT` wraps nothing.

mhaessig replied:
The initialization list should only be generated if `!PRODUCT && ASSERT`, so it does not appear in `optimized`. This is one way of expressing this conjunction in macros.

chhagedorn commented on src/hotspot/os/linux/globals_linux.hpp:
```diff
@@ -89,6 +89,11 @@
                                                                         \
+  develop(intx, CompileTaskTimeout, 0,                                  \
+          "Set the timeout for compile tasks' CPU time in milliseconds."\
+          "0 = no timeout (default)")                                   \
```
```suggestion
          " 0 = no timeout (default)")                                  \
```

chhagedorn commented on src/hotspot/share/compiler/compilerThread.hpp:
```diff
@@ -38,10 +45,25 @@ class CompileQueue;
 class CompilerCounters;
 class IdealGraphPrinter;
 
+class CompilerThreadTimeoutGeneric : public CHeapObj<mtCompiler> {
+ public:
+  CompilerThreadTimeoutGeneric() {};
+  void arm() { return; };
+  void disarm() { return; };
+  bool init_timeout() { return true; };
+};
```
Should we also guard this with `ifndef LINUX` since it's only used for non-Linux?

mhaessig replied:
Sounds reasonable.

chhagedorn commented on src/hotspot/share/compiler/compilerThread.hpp:
```diff
@@ -38,10 +45,25 @@ class CompileQueue;
+class CompilerThreadTimeoutGeneric : public CHeapObj<mtCompiler> {
+ public:
+  CompilerThreadTimeoutGeneric() {};
+  void arm() { return; };
+  void disarm() { return; };
```
You can remove the `return`:
```suggestion
  void arm() {};
  void disarm() {};
```

chhagedorn reviewed (commented):
Nice improvement! I left some small comments in the code but otherwise the change looks reasonable!

Can we also add some tests for the new `CompileTaskTimeout` flag? Maybe we can add a positive test and negative test:
- Positive test: Could just be a hello world test with a reasonably large non-zero value for `CompileTaskTimeout`.
- Negative test: Maybe we can just set `CompileTaskTimeout=1` which will probably crash immediately for a hello world program. That could be run in a separate VM and then we can check the output. If we are able to also dump the compile task/method that is timing out, we might even be able to match on that when run with `CompileOnly` for a single method. But not sure if the latter is possible.

What do you think?

chhagedorn commented on test/hotspot/jtreg/compiler/arguments/TestCompileTaskTimeout.java:
```diff
@@ -0,0 +1,48 @@
+/*
+ * @test TestCompileTaskTimeout
+ * @bug 8308094
+ * @requires vm.compiler2.enabled & vm.debug & vm.flagless & os.name == "Linux"
```
Does it only work with C2 compile tasks?

chhagedorn commented on src/hotspot/os/linux/compilerThreadTimeout_linux.cpp:
```diff
@@ -40,15 +39,16 @@ void compiler_signal_handler(int signo, siginfo_t* info, void* context) {
     case TIMEOUT_SIGNAL: {
-      assert(false, "compile task timed out");
+      CompileTask* task = CompilerThread::current()->task();
+      assert(false, "compile task %d (%s) timed out after " INTPTR_FORMAT " ms",
+             task->compile_id(), task->method()->name_and_sig_as_C_string(), CompileTaskTimeout);
```
Normally, you would probably need a `ResourceMark` for getting the method name. However, I'm not sure if you can do that as well here inside the signal handler. Maybe someone else can comment on that.

chhagedorn commented on test/hotspot/jtreg/compiler/arguments/TestCompileTaskTimeout.java:
```diff
@@ -0,0 +1,48 @@
+public class TestCompileTaskTimeout {
+
+    public static void main(String[] args) throws Throwable {
+        ProcessTools.executeTestJava("-Xcomp", "-XX:CompileTaskTimeout=1", "-version")
```
Nit: I think for newer JDK versions after JDK 8, we should use `--version`.

chhagedorn reviewed (approved):
Thanks a lot for adding a test and printing the method name! A few more comments but otherwise, it looks good now!

dean-long commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -151,6 +151,7 @@ class oopDesc;
 #endif
 
 // Format pointers and padded integral values which change size between 32- and 64-bit.
+#define INTX_FORMAT              "%"          PRIdPTR
```
Do we really need to add this back?  This was removed by JDK-8346990.
